app/models/automation_worker.rb

```diff
@@ -0,0 +1,13 @@
+class AutomationWorker < MiqQueueWorkerBase
+  include MiqWorker::ReplicaPerWorker
+
+  require_nested :Runner
+
+  self.required_roles               = ["automate"]
```

Comment:
From a migration perspective:

I had thought we would increase the number of workflow workers and drop the automate role where appropriate. But since I guess there are no automate workers per se, that may be moot.

From a resources perspective:

I like the way the event handler workflow type works. Can we duplicate that? Does the event handler rely upon a `server_guid`? (avoiding server_guid would be nice, but if that is what the event handler uses...)

We did try to have an automate worker per server. But memory requirements forced us to rollback dedicated automate workers and go back to generalized automate workers. This conversation feels similar.

Reply:
> I like the way the event handler workflow type works. Can we duplicate that? 

That has `max_concurrent=1` and `role_scope=zone`

>  Does the event handler rely upon a server_guid? (avoiding server_guid would be nice, but if that is what the event handler uses...)

No, but a. the event handler is only active on one miq_server and b. the event_handler doesn't have any server-local resources like running podman containers

Reply:
I was also worried about the automate role - think I mentioned that elsewhere.  I thought someone said this would be handled by queue_name?  My only concern is this will start consuming work that it's not quite ready for (i.e. embedded automate stuff).

> I initially pictured this as maybe one worker per zone running all of the workflows rather than one-per-miq_server.

I was picturing it as 1 maximum per miq_server where the role is configured.  What I want to avoid is us preventing scaling, and one per zone would prevent scaling within that zone because then it would only run on one server and consume that server's local resources.  1 per miq_server should still allow zone _placement_, as long as the role is configured in that zone, so you can still get benefits if the user has the zone network configuration in a particular way. (That being said, I don't think we expose which zone a workflow can run on so that's a future feature right now).

The other side effect of 1 per miq_server is that kubernetes is considered one giant server, so on kube we would only have 1 worker max talking to kube which keeps the security footprint small, while still having the scalability of the entire cluster.

> I like the way the event handler workflow type works.

Interestingly, I _don't_ like how that one works, so it's not a pattern I'd like to follow.  It forces a single worker to do everything, serializing all event handling, and doesn't support scaling.  We require serialization of event handling because you don't want to process a "complete" event before a "started" event, but that serialization is only really needed per-provider, so I'd like to see this worker become 1-max-per-provider.

Reply:
if we have one per provider, I don't think this necessarily forces that they are processed sequentially. after you queue event A, B, C

I can see something like this happening for different tasks. (the same type will complete in order A, B, C, but intermittent processing is possible:

1. process event A. it goes out to a task/blocks so it pushes it back on the queue
2. process event B. it gets blocked, pushes onto queue
3. process event C. it completes and is done
4. process event A. it does stuff but then blocks again
5. process event B. completes
6. process event A. completes

Reply:
@Fryguy I think there is a different concern

The generic worker will not get items targeting the workflow worker or vice versa. I agree with adam here.

However the problem is that the old automate, old embedded ansible, and new workflow worker potentially have different role scope strategies.

Reply:
I think we're going with one per miq-server with the automate role, I'm not sure we guarantee ordering of processing of different workflows, only ordering within a workflow.

Reply:
>  I thought someone said this would be handled by queue_name? My only concern is this will start consuming work that it's not quite ready for (i.e. embedded automate stuff).

Correct, embedded_automate uses a `nil` queue_name which means it'll be process by priority/generic workers and not this worker: https://github.com/ManageIQ/manageiq-automation_engine/blob/master/lib/miq_automation_engine/engine/miq_ae_engine.rb#L18-L25 same with embedded_ansible https://github.com/ManageIQ/manageiq/blob/master/app/models/service_ansible_playbook.rb#L35-L42

Reply:
Sorry I didn't want to derail this thread with discussion of the event handler, but the event handler ordering is guaranteed simply because we don't "push it back onto the queue".